Assets/PatchKit Patcher/Scripts/AppData/Remote/Downloaders/TorrentDownloader.cs

```diff
@@ -67,47 +105,25 @@ public void Download(CancellationToken cancellationToken)
 
                 _logger.LogTrace("download dir  = " + DestinationDirectoryPath);
 
-                using (var torrentClient = new TorrentClient(new UnityTorrentClientProcessStartInfoProvider()))
-                {
-                    torrentClient.AddTorrent(_torrentFilePath, DestinationDirectoryPath, cancellationToken);
-
-                    var timeoutWatch = new Stopwatch();
-                    timeoutWatch.Start();
-
-                    TorrentStatus status = GetAndCheckTorrentStatus(torrentClient, cancellationToken);
-                    double initialProgress = status.Progress;
-                    _logger.LogTrace("initialProgress = " + status.Progress);
-                    var waitHandle = new AutoResetEvent(false);
+                IRequestRetryStrategy retryStrategy = new RetryStrategy(3);
 
-                    OnDownloadProgressChanged(0);
-
-                    using (cancellationToken.Register(() => waitHandle.Set()))
+                do
+                {
+                    try
                     {
-                        bool finished = false;
+                        DownloadInternal(cancellationToken);
+                    }
+                    catch (TorrentClientException e)
+                    {
+                        _logger.LogError("Exception in torrent client.", e);
+                        retryStrategy.OnRequestFailure();
 
-                        do
+                        if (!retryStrategy.ShouldRetry)
                         {
-                            cancellationToken.ThrowIfCancellationRequested();
-
-                            status = GetAndCheckTorrentStatus(torrentClient, cancellationToken);
-
-                            _logger.LogTrace("progress = " + status.Progress);
-
-                            CheckTimeout(timeoutWatch, status.Progress, initialProgress);
-
-                            OnDownloadProgressChanged((long) (_totalBytes * status.Progress));
-
-                            if (status.IsSeeding)
-                            {
-                                finished = true;
-                            }
-                            else
-                            {
-                                waitHandle.WaitOne(UpdateInterval);
-                            }
-                        } while (!finished);
+                            throw new DownloadFailureException("Failed to download with torrent-client.");
```

Comment:
To keep the error information we should put `e` as inner exception inside `DownloadFailureException`.